lib/category_extension.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+class Category
+  module GlobalFilter::CategoryExtension
+    def update_stats
+      global_filters = SiteSetting.find_by(name: "global_filters").value.split("|")
+      topics_with_post_count = Topic
+        .select("topics.category_id, COUNT(*) topic_count, SUM(topics.posts_count) post_count")
+        .where("topics.id NOT IN (SELECT cc.topic_id FROM categories cc WHERE topic_id IS NOT NULL)")
+        .where(<<~SQL, global_filters: global_filters)
+          topics.id IN (
+            SELECT topic_tags.topic_id
+            FROM topic_tags INNER JOIN tags
+            ON tags.id = topic_tags.tag_id
+            WHERE tags.name IN (:global_filters)
+          )
+        SQL
+        .group("topics.category_id")
+        .visible.to_sql
+
+      DB.exec <<~SQL
+        UPDATE categories c
+           SET topic_count = COALESCE(x.topic_count, 0),
+               post_count = COALESCE(x.post_count, 0)
+          FROM (
+                SELECT ccc.id as category_id, stats.topic_count, stats.post_count
+                FROM categories ccc
+                LEFT JOIN (#{topics_with_post_count}) stats
+                ON stats.category_id = ccc.id
+               ) x
+         WHERE x.category_id = c.id
+           AND (c.topic_count <> COALESCE(x.topic_count, 0) OR c.post_count <> COALESCE(x.post_count, 0))
+      SQL
+
+      Category.all.each do |c|
```

Comment:
Just in case, don't preload **all** categories at once:

```suggestion
      Category.find_each do |c|
```